### PR TITLE
[core] Introduce binlog system table to pack the UB and UA

### LINKED ISSUE #4505: [Feature] Allow to pack the UPDATE_BEFORE and UPDATE_AFTER message when reading the audit table
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Motivation

Flink breaks the ub and ua into two records, however, user may want to compare the ub and ua message in the udf to find the diff column. So, I proposal to add a mode in aduit_table to support consume the ub and ua message in one row. 

We could make the original column type to an array type.

for example:

```
    public void testAuditTablePackUbUa() throws Exception {
        sql(
                "CREATE TABLE T (a INT, b INT, primary key (a) NOT ENFORCED) with ('changelog-producer' = 'lookup', 'bucket' = '2', 'scan.pack.update-before-after.enabled' = 'true')");
        BlockingIterator<Row, Row> iterator =
                streamSqlBlockIter(
                        "SELECT * FROM T$audit_log /*+ OPTIONS('scan.mode' = 'latest') */");
        sql("INSERT INTO T VALUES (1, 2)");
        sql("INSERT INTO T VALUES (1, 3)");
        sql("INSERT INTO T VALUES (2, 2)");
        List<Row> rows = iterator.collect(3);
        System.out.println(rows);
        iterator.close();
    }
```

This will produce 

```
[+I[+I, [1, null], [2, null]], +I[+U, [1, 1], [2, 3]], +I[+I, [2, null], [2, null]]]
```

If the column has two index, then the first index is update_before value, and the second index is update_after value. 

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Aitozi**
### Purpose

Linked issue: close #4505

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
I feel like we can create a new system table instead of an option.

**Comment by Aitozi** [comment]
> I feel like we can create a new system table instead of an option.

Yes, it's better. Updated.

File: docs/content/maintenance/system-tables.md [hunk 1]
```diff
@@ -406,4 +406,23 @@ SELECT * FROM T$statistics;
...
+
+You can streaming or batch query the binlog through binlog table. In this system table,
+the update before and update after will be packed in one row. 
+
+```
+/*
++------------------+----------------------+-----------------------+
+|     rowkind      |     column_0         |     column_1          |
++------------------+----------------------+-----------------------+
+|        +I        |  [col_0, null]       |   [col_1, null]       |
```
Comment (JingsongLi) [inline docs/content/maintenance/system-tables.md:419]: Why not just one element?

File: paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+/**
+ * A {@link Table} for reading binlog of table. The binlog format is as below.
+ *
+ * <p>INSERT: [+I, [co1, null], [col2, null]]
+ *
+ * <p>UPDATE: [+U, [co1_ub, col1_ua], [col2_ub, col2_ua]]
+ *
+ * <p>DELETE: [-D, [co1, null], [col2, null]]
+ */
+public class BinlogTable implements DataTable, ReadonlyTable {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java:96]: Can we reuse code with `AuditLogTable`? Like extending it to reuse something.

File: paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+
+    public static final String BINLOG = "binlog";
+
+    public static final PredicateReplaceVisitor PREDICATE_CONVERTER =
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java:100]: private

File: paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+    }
+
+    @Override
+    public String name() {
+        return wrapped.name() + SYSTEM_TABLE_SPLITTER + BINLOG;
+    }
+
+    @Override
+    public RowType rowType() {
+        List<DataField> fields = new ArrayList<>();
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java:152]: init map with capacity

File: paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java (hunk 2 above)
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/system/BinlogTable.java:96]: May be can extract some common code with the AuditLogTable system table.

File: paimon-flink/paimon-flink-common/pom.xml [hunk 5]
```diff
@@ -162,6 +162,12 @@ under the License.
             <artifactId>iceberg-data</artifactId>
             <version>${iceberg.version}</version>
             <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <artifactId>parquet-avro</artifactId>
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/pom.xml:167]: Why do this here?
It seems has no relation with this PR?

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/CatalogTableITCase.java [hunk 6]
```diff
@@ -132,6 +132,40 @@ public void testSnapshotsTable() throws Exception {
                         Row.of(3L, 0L, "APPEND"));
     }
 
+    @Test
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/CatalogTableITCase.java:135]: I think add a class of BinlogTableTest in package "org.apache.paimon.table.system" may better?

File: docs/content/maintenance/system-tables.md (hunk 1 above)
Comment (Aitozi) [reply docs/content/maintenance/system-tables.md:419]: The schema should keep same, so this is an array only with one element.

File: paimon-flink/paimon-flink-common/pom.xml (hunk 5 above)
Comment (Aitozi) [reply paimon-flink/paimon-flink-common/pom.xml:167]: See the first commit message

File: paimon-common/src/main/java/org/apache/paimon/reader/PackChangelogReader.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        @Nullable
+        @Override
+        public InternalRow next() throws IOException {
+            InternalRow row1 = nextRow();
+            if (row1 == null) {
+                return null;
+            }
+            InternalRow row2 = null;
+            if (row1.getRowKind() == RowKind.UPDATE_BEFORE) {
+                row1 = serializer.copy(row1);
```
[the remark this answers is not in the document]
Comment (Aitozi) [reply paimon-common/src/main/java/org/apache/paimon/reader/PackChangelogReader.java:91]: The underlay's row fetched from the `nextRow` could be reused.

File: docs/content/maintenance/system-tables.md (hunk 1 above)
Comment (JingsongLi) [reply docs/content/maintenance/system-tables.md:419]: I mean one element, not one element and a null, two elements.

Comment (Aitozi) [reply docs/content/maintenance/system-tables.md]: Get it, updated.

**Comment by Aitozi** [comment]
Addressed the comments, pls take a look again @JingsongLi @wwj6591812

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**